signon-plugin-oauth2: Don't send code_challenge when token is requested

Thaodan opened the pull request:
In case a service sends token as a response type imply that PKCE shall
not be used.

Also rebase patches with  format-patch --no-numbered --zero-commit --no-signature --full-inde.

Tomin1 commented on rpm/0007-Add-RFC7636-aka-PKCE-support.patch:
```diff
@@ -110,7 +111,7 @@ index a5f8e9769939ad1e0fe04840ef9208365167369b..452d03a88e927a26a5860a572efe98ef
      }
 +    // PKCE (RFC7636)
-+    if (!d->m_oauth2Data.SkipPKCE()) {
++    if (!d->m_oauth2Data.SkipPKCE() &&!responseType.contains("token")) {
```
I think this should be 
```C++
if (!d->m_oauth2Data.SkipPKCE() && responseType.contains("code")) {
```
since [RFC 7636 section 4.3](https://www.rfc-editor.org/rfc/rfc7636#section-4.3) defines that the code is sent with "Authorization Request" as the only situation when it would be sent.

I guess if some extensions want to use it, then they would need changes of course but that  support needs to be added on top of the RFCs anyway.

Also pondering a bit if `"code"` should be a constant at the beginning of the file, but that's not that important. There is already `AUTH_CODE` there but that's for a different purpose and I'm not sure how I'd feel about (ab)using it.

Thaodan replied:
>  Also pondering a bit if "code" should be a constant at the beginning of the file, but that's not that important. 
> There is already AUTH_CODE there but that's for a different purpose and I'm not sure how I'd feel about (ab)using it.

I checked the AUTH_CODE constant it is already used in similar use cases, however I'm not sure
if it makes sense to do such refactoring alone. 
I would ask upstream when posting these patches.

Tomin1 replied:
Please don't get stuck with that `"code"`, it's fine as a string or reusing that `AUTH_CODE`. I will approve it either way, that first part of my comment is much more important. :slightly_smiling_face:

Tomin1 commented on rpm/0007-Add-RFC7636-aka-PKCE-support.patch:
```diff
@@ -22,18 +22,19 @@ Linux but might behave differently on other platforms and is not
+@@ -120,6 +120,16 @@ namespace OAuth2PluginNS {
           */
-         SIGNON_SESSION_DECLARE_PROPERTY(QString, TokenQuery);
+         SIGNON_SESSION_DECLARE_PROPERTY(QString, AuthQuery);
```
I think this part of the change shouldn't be necessary. TokenQuery could just as well stay before these properties.

Tomin1 reviewed (commented):
One important comment to address and one that's only cosmetic (so it's more up to you then).

Thaodan commented on rpm/0007-Add-RFC7636-aka-PKCE-support.patch:
```diff
@@ -22,18 +22,19 @@ Linux but might behave differently on other platforms and is not
           */
-         SIGNON_SESSION_DECLARE_PROPERTY(QString, TokenQuery);
+         SIGNON_SESSION_DECLARE_PROPERTY(QString, AuthQuery);
  
```
The change is from patch -p1 -i after git am failed, I can drop it.

Tomin1 replied:
Personally I would try to keep the commits as they were so these wouldn't move around.

Thaodan replied:
Without this change the patches don't apply using git am, the change is from git format-patch not
from me.

Tomin1 replied:
Actually, I think there is a small mistake in patch 0006. It is missing a newline after that TokenQuery and that's why these don't quite match. I wonder how I made that mistake.

```
-@@ -124,6 +124,10 @@ namespace OAuth2PluginNS {
+@@ -124,6 +124,11 @@ namespace OAuth2PluginNS {
           * query component of the token endpoint URI
           */
          SIGNON_SESSION_DECLARE_PROPERTY(QString, TokenQuery);
++
 +        /*!
 +         * Any extra (non standard) parameters to send to the endpoint URI
 +         */
```

After adding that newline 0007 applies.

Thaodan replied:
Oh that explains it, I fixed the 0006 patch now and rebased 0007 after

Tomin1 commented on rpm/0007-Add-RFC7636-aka-PKCE-support.patch:
```diff
@@ -110,7 +111,7 @@ index a5f8e9769939ad1e0fe04840ef9208365167369b..452d03a88e927a26a5860a572efe98ef
      }
 +    // PKCE (RFC7636)
-+    if (!d->m_oauth2Data.SkipPKCE()) {
++    if (!d->m_oauth2Data.SkipPKCE() &&responseType.contains("code")) {
```
A space after `&&`, please. Otherwise this looks good. I'll try it myself too.

Thaodan commented on rpm/0007-Add-RFC7636-aka-PKCE-support.patch:
```diff
@@ -22,18 +22,19 @@ Linux but might behave differently on other platforms and is not
           */
-         SIGNON_SESSION_DECLARE_PROPERTY(QString, TokenQuery);
+         SIGNON_SESSION_DECLARE_PROPERTY(QString, AuthQuery);
  
```
Without this change the patches don't apply using git am, the change is from git format-patch not
 from me.

Thaodan committed:
signon-plugin-oauth2: Only send code_challenge when code is requested

In case a service sends token as a response type imply that PKCE shall
not be used.

Also rebase patches with  format-patch --no-numbered --zero-commit --no-signature --full-inde.

[signon-plugin-oauth2] Only send code_challenge when code is requested. Fixes JB#59112

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan commented:
I wonder if we should update the description of the 7th patch since the code challenge is only send when the response type was code.

Tomin1 commented:
> I wonder if we should update the description of the 7th patch since the code challenge is only send when the response type was code.

Sure. Feel free to update it. :)

Tomin1 reviewed (approved):
LGTM! Nice work!
